src/components/tl/ucp/allreduce/allreduce_sliding_window.c

```diff
@@ -0,0 +1,323 @@
+        tl_error(UCC_TASK_LIB(task), "unable to complete UCX request=%p: %d\n",
+                 request,
+                 UCS_PTR_STATUS(request));
+        return UCS_PTR_STATUS(request);
+    } else {
+        return ucp_request_check_status(request);
+    }
+}
+
+void ucc_tl_ucp_allreduce_sliding_window_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t                  *task           = ucc_derived_of(coll_task,
+                                                                        ucc_tl_ucp_task_t);
+    ucc_rank_t                          size           = (ucc_rank_t)task->subset.map.ep_num;
+    ucc_datatype_t                      dtype          = TASK_ARGS(task).dst.info.datatype;
+    size_t                              dt_size        = ucc_dt_size(dtype);
+    uint32_t                            host_team_size = size;
+    ucc_tl_ucp_allreduce_sw_pipeline_t *pipe           = task->allreduce_sliding_window.pipe;
+    ucc_tl_ucp_allreduce_sw_buf_t      *accbuf         = &pipe->accbuf;
+    ucp_request_param_t                 req_param      = {0};
+    int                                 i              = 0;
+
+    size_t                         remaining_elems;
+    size_t                         get_idx;
+    size_t                         count;
+    size_t                         get_offset;
+    size_t                         data_size;
+    int                            src_rank;
+    void                          *src_addr;
+    void                          *dst_addr;
+    ucs_status_ptr_t               request;
+    size_t                         red_idx;
+    ucc_tl_ucp_allreduce_sw_buf_t *redbuf;
+    ucc_tl_ucp_allreduce_sw_buf_t *getbuf;
+    size_t                         put_offset;
+    int                            window;
+    int                            put_idx;
+    int                            dst_rank;
+
+    if (pipe->count_serviced < pipe->my_count) {
+        if ((pipe->count_received < pipe->my_count)
+            && (pipe->done_get < host_team_size)
+            && (pipe->avail_buffs > 0)
+            && (accbuf->state != REDUCED && accbuf->state != SENDING))
+        {
+            remaining_elems = pipe->my_count - pipe->count_received;
+            get_idx         = pipe->get_idx % pipe->num_buffers;
+            count           = ucc_min(pipe->buffer_size/dt_size,
+                                      remaining_elems);
+            get_offset      = pipe->count_received * dt_size
+                                + pipe->my_offset;
+            data_size       = count * dt_size;
+            src_rank        = pipe->src_rank;
+            getbuf          = accbuf->state == FREE ? 
+                                accbuf : &pipe->getbuf[get_idx];
+            src_addr        = task->allreduce_sliding_window.sbufs[src_rank]
+                                + get_offset;
+            dst_addr        = getbuf->buf;
+
+            assert(getbuf->state == FREE);
```

Comment:
Is there a way to ensure that the buffer is free ? I see that `pipe->avail_buffs > 0` but is it sufficient to ensure that this precise buffer is free ?

Reply:
If the get buf is not free, it is because it is either recv'ing or reducing. It cannot be recv'ing, because avail_bufs > 0 and get_idx is incremented after the buffer is set to recv'ing. It cannot be reducing, because red_idx is always behind get_idx.